flang/lib/Lower/Bridge.cpp

```diff
@@ -3224,8 +3224,13 @@ class FirConverter : public Fortran::lower::AbstractConverter {
       mlir::Location loc, const Fortran::evaluate::Assignment &assign,
       const Fortran::evaluate::Assignment::BoundsSpec &lbExprs) {
     Fortran::lower::StatementContext stmtCtx;
-    if (Fortran::evaluate::IsProcedure(assign.rhs))
-      TODO(loc, "procedure pointer assignment");
+
+    if (Fortran::evaluate::IsProcedure(assign.rhs)) {
+      auto lhs{fir::getBase(genExprAddr(assign.lhs, stmtCtx, &loc))};
+      auto rhs{fir::getBase(genExprAddr(assign.rhs, stmtCtx, &loc))};
```

Comment:
Thanks for the comments, and it is a good point! I am planning to have a separate PR to handle procedure pointer argument and function results. I will keep this suggestion in mind when I work on that.
As for this PR, the target is external, module or internal procedure.

Reply:
Another point is that you may need custom handling of `Fortran::evaluate::UnwrapExpr<Fortran::evaluate::NullPointer>(assign.rhs)` like it is done for data pointer assignments since `Fortran::evaluate::NullPointer` is untyped and context dependent (I do not think it will be lowered to a fir.boxproc zero initialized).

Reply:
I adopted your suggestion to have
```
    if (Fortran::evaluate::IsProcedurePointerTarget(assign.rhs)) {
      hlfir::Entity lhs = Fortran::lower::convertExprToHLFIR(
          loc, *this, assign.lhs, localSymbols, stmtCtx);
      hlfir::Entity rhs = Fortran::lower::convertExprToHLFIR(
          loc, *this, assign.rhs, localSymbols, stmtCtx);
      rhs = hlfir::derefPointersAndAllocatables(loc, *builder, rhs);
      builder->createStoreWithConvert(loc, rhs, lhs);
      return;
    }
    ```
    
    I ran into the following issue when do `pptr => foo()` where `foo` returns a procedure pointer.
    ```
    flang/lib/Lower/ConvertExprToHLFIR.cpp:1428: not yet implemented: lowering ProcRef to HLFIR
    ```
    Is lowering ProcRef being worked on?

Reply:
Thanks for the update! No one is working on this TODO (quite frankly, I thought it would not be hit, it skipped my mind that functions returning procedure pointers are not implemented as evaluate::Function<T>). I suggest rephrasing it to "lowering function references that return procedure pointers" and not try to implement it in this PR.

I think you have enough content for a first patch and that functions returning procedure pointers can be dealt in another patch. The main thing now is to ensure that it is covering everything case for the code path where the TODO have been removed, and if not, to add more narrow TODOs for the leftover.

Reply:
Ok. I will update the TODO message as suggested. I will not include any form of referencing a function that returns a procedure pointer in this PR then. Thanks for confirming it!